### PR TITLE
Single Braid Campaign Contract and Tests

### PR DESCRIPTION
**Pull request by SilentCicero**
## weifund-braid-contract

Issue:
Our current campaign contracts do not take into the consideration a fluctuating token price, and that the beneficiary must be paid out every time a contribution is made.

This is paramount to the Braid campaign and their requirements.

I created a backup contract, incase we want to switch to something that does only what we need it to do, vs something generic in design.

Solution:
A single contract, the campaign is the token contract, price can be changed by the beneficiary, the campaign can be stopped by the beneficiary.

Follow, to a T, every existing interface so that it works with existing JS. Not a line of JS should need to be changed. However, less throws should occur.

Follow the same verifier model/API, so no server code should be changed.

Simple two state contract, no refunds and payouts are required. No claim of tokens is also required.

Tokens are issued automatically, but can only be transferred and approved after the freeze period is over. 

Features (list format):
- Single campaign contract
- Simple design
- Identical interface to existing contracts
- No JS should need to be changed (including server/front-end)
- Two states (operational and success)
- Beneficiary can trigger a stop at any time (if compromised)
- Verifier api/code remains the same
- Less throws should occur at the contribution amount can be above or below price of token
- Only token cap and expiry is enforced (no ether cap)
- No claim process (tokens are auto issued at contribution)
- Allows fallback function to be used (if they are approved)

Considerations:
- We may want the beneficiary to only receive funds, and the "owner" to be the one to change price and stop the campaign (if there are problems)
- Note, this would be an easy fix/addition
- Price target is at 1 USD, which means a max of 1 USD of current ether losses if the contribution is not exact units of the 1 USD price in wei. Not horrible, but some ether may be lost in contributions.

Tests:
- I wrote a series of tests covering all scenarios I could think of:
1. TestBraidCampaign_deployment --- deployment checks
2. TestBraidCampaign_contribution --- contributions function normally
3. TestBraidCampaign_earlySuccess --- early success trigegrs properly
4. TestBraidCampaign_changePrice --- change price triggers properly
5. TestBraidCampaign_capReach --- cap reach scenario functions
6. TestBraidCampaign_expiryReach --- expiry reach scenario functions
7. TestBraidCampaign_fundingCap_edgeCase --- covers when eth amount is non-price unit, close to cap reach
8. TestBraidCampaign_fullCampaign_withSuccessTransfers --- covers campaign in full swing (price changes, multiple contributors, cap reach, post token transfers by contributors after freeze period)

Review:
- Both J. Chow and Aakil have given their feedback on the contract. Will be posting in a quip doc today.

### CODE REVIEW

File: src/contracts/campaigns/BraidCampaign.sol [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+import "tokens/StandardToken.sol";
+
+/*
+A custom WeiFund compatable campaign for the Braid the Movie Ethereum campaign.
+*/
+
+contract BraidCampaign is Owned, StandardToken {
+  modifier pastFreezeDate {
+    // now must be greater than freeze date
+    if (block.number >= freezePeriod) {
```
Comment (niran) [inline src/contracts/campaigns/BraidCampaign.sol:15]: `freezePeriod` is still a confusing name because it sounds like a count of blocks rather than a block number. `freezeUntilBlock`?

File: src/contracts/campaigns/BraidCampaign.sol [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  // default amounts used
+  uint256[] defaultAmounts;
+
+  // the stop param, which can be triggered by the beneficiary to stop the campaign
+  bool public earlySuccess;
+
+  // campaign enhancer, usually for token notation
+  address public token;
```
Comment (niran) [inline src/contracts/campaigns/BraidCampaign.sol:136]: `token` isn't used anymore.

File: src/contracts/campaigns/BraidCampaign.sol [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+  // campaign enhancer, usually for token notation
+  address public enhancer;
```
Comment (niran) [inline src/contracts/campaigns/BraidCampaign.sol:139]: `enhancer` isn't used anymore.

File: src/contracts/campaigns/BraidCampaign.sol [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  function () payable public {
+    // allow fallback function.. contribute as normal
+    contributeMsgValue(defaultAmounts);
+  }
+
+  function contributeMsgValue(uint256[] _amounts) payable public returns (uint256) {
+    // the total tokens amount to be issued to the sender
+    uint256 tokensOwed = msg.value / price;
```
Comment (niran) [inline src/contracts/campaigns/BraidCampaign.sol:41]: This no longer throws when there's remaining ether that doesn't by any tokens, so ether gets thrown away. The only valid options are throwing or refunding the extra ether. We can't just silently keep it without granting any tokens for it.

**Review by niran — CHANGES_REQUESTED** [review]

File: src/contracts/campaigns/BraidCampaign.sol (hunk 1 above)
Comment (SilentCicero) [reply src/contracts/campaigns/BraidCampaign.sol:15]: Yes, the only reason I did this was to keep the interface the same. We can change, but JS will have to be changed also.

File: src/contracts/campaigns/BraidCampaign.sol (hunk 2 above)
Comment (SilentCicero) [reply src/contracts/campaigns/BraidCampaign.sol:136]: It's used by the JS, and my monitor.

File: src/contracts/campaigns/BraidCampaign.sol (hunk 3 above)
Comment (SilentCicero) [reply src/contracts/campaigns/BraidCampaign.sol:139]: This is pointing to itself, so that my monitor and other tools will still work.

File: src/contracts/campaigns/BraidCampaign.sol (hunk 4 above)
Comment (SilentCicero) [reply src/contracts/campaigns/BraidCampaign.sol:41]: The client wants to grant token to have more precision than this. I just did this as this code above, because it's our current price calculation. I'm talking with them now. He says he wants tokens that are divisible and have more "decimals" than this. So I'm assuming we allow more precision from the msg.value number, thus the contributor would not loose nearly as much based on the contribution amount.

So we just need to work up a scheme with more precision, and make sure the contributor gets more of their contributions worth of tokens given the USD price.

Comment (niran) [reply src/contracts/campaigns/BraidCampaign.sol:41]: If we want more resolution, we can make tokens have 1e18 units, just like ether. (Let's call those units atoms to make it easier to talk about. One token has 1e18 atoms.) We need to go from wei to atoms without losing precision.

Currently, the price set by `payout-server` is in wei per token. If we want wei per atom, we need to divide the price by 1e18. If we do that, we're going to get zero—each token costs less than 1e18 wei, so integer division will discard the remainder. Instead, let's get the price in wei per 1e9 atoms. We're trying to avoid losing cents, and 1e9 atoms is one-ten millionth of a cent, so it's okay to lose precision there.

The price in wei per 1e9 atoms is `1e18 / usdPerEther / 1e9`. At $50 per ether, for instance, the price would be 2e7 wei per 1e9 atoms. In the contract, we'd take `msg.value / price * 1e9` to get the number of atoms the user should be issued for the ether they sent. A `decimals` attribute needs to be added to the contract and set to 18 so ERC20-compliant tools know to divide the value by 1e18 to display the number of tokens a user has.

TL;DR: Change `payout-server` to set the price as `const weiPricePerBillion = oneEther.div(bodyData.mid).div('1e9');`, then change the contract to set `uint256 tokensOwed = msg.value / price * 1e9;` with a comment that reminds the reader that `price` is in ETH wei per billion token wei.

**Review by niran — APPROVED** [review]

**Comment by niran** [comment]
The only outstanding issue is the precision issue that isn't related to the original pull request, so this is good to go. I still think using new contracts for this campaign is a bad idea.

File: src/contracts/campaigns/BraidCampaign.sol (hunk 4 above)
Comment (SilentCicero) [reply src/contracts/campaigns/BraidCampaign.sol:41]: Thanks @niran I will make changes, and update the tests and send to the auditing teams.

**Comment by SilentCicero** [comment]
@niran I understand the concern, it does feel rushed, and doing the/a bug bounty a second time may get less attention than before. However, given the simplicity of the contracts and enough oversight I think that + a small second bug bounty should be adequate. Let's do a full review Monday.

File: src/contracts/campaigns/BraidCampaign.sol (hunk 4 above)
Comment (SilentCicero) [reply src/contracts/campaigns/BraidCampaign.sol:41]: @niran so with this calculation, would we be only throwing away less than a cent (at current prices)?
